ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,271 @@
+    )
+"""
+
+DEFAULT_CACHE_SIZE = 100000
+
+
+class Node(NamedTuple):
+    id: int
+    key: Union[str, bytes]
+    value: Any
+    pref: int
+    nref: int
+
+
+class LRUSQLDict(MutableMapping[Union[str, bytes], Any]):
+    """
+    SQLite3-backed dictionary that implements an LRU cache.
+    """
+
+    def __init__(
+        self, conn: sqlite3.Connection, cache_size: int = DEFAULT_CACHE_SIZE
+    ) -> None:
+        self._conn = conn
+        self._execute(CREATE_CACHE_QUERY)
+        self._execute(CREATE_MAP_QUERY)
+        self.cache_size = cache_size
+
+    def __iter__(self) -> Iterator[Union[str, bytes]]:
+        with self._conn:
+            for key in self._conn.execute("SELECT key FROM cache"):
+                yield first(key)
+
+    def __len__(self) -> int:
+        result = self._fetch_single_query("SELECT COUNT(*) FROM cache;")
+        return int(first(result))
+
+    def __setitem__(self, key: Union[str, bytes], value: Any) -> None:
+        # updates LRU cache
+        try:
+            self.__getitem__(key)
+        except KeyError:
+            self._insert_item(key, value)
+        else:
+            self._update_item(key, value)
+
+    def __getitem__(self, key: Union[str, bytes]) -> Any:
+        # updates LRU cache
+        pointer_result = self._fetch_single_query(
+            "SELECT pointer FROM map WHERE key=?;", (key,),
+        )
+
+        if not pointer_result:
+            raise KeyError(f"Key ({str(key)}) not found in db.")
+
+        (map_pointer,) = pointer_result
+
+        lookup_result = self._fetch_single_query(
+            "SELECT value FROM cache WHERE id=?;", (map_pointer,),
+        )
+
+        (value,) = lookup_result
+
+        unpickled_value = self._unpickle(value)
+
+        # update cache
+        self.__delitem__(key)
+        self._insert_item(key, unpickled_value)
+
+        return unpickled_value
+
+    def __delitem__(self, key: Union[str, bytes]) -> None:
+        result = self._fetch_single_query(
+            "SELECT id,key,value,pref,nref FROM cache WHERE key=?;", (key,),
+        )
+
+        if not result:
+            raise KeyError(f"Cannot delete key. Key: {str(key)} not found in db.")
+
+        node = Node(*result)
+
+        # delete key from cache & map
+        self._execute(
+            "DELETE FROM map WHERE pointer=?;", (node.id,),
+        )
+        self._execute(
+            "DELETE FROM cache WHERE id=?;", (node.id,),
+        )
+
+        # update any nrefs/prefs in cache
+        nref_node = self._fetch_single_query(
+            "SELECT id,key,value,pref,nref FROM cache WHERE nref=?;", (node.id,),
+        )
+        pref_node = self._fetch_single_query(
+            "SELECT id,key,value,pref,nref FROM cache WHERE pref=?;", (node.id,),
+        )
+
+        if nref_node and pref_node:
+            nref = Node(*nref_node)
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE id=?;", (pref.id, nref.id),
```

Comment:
You should be able to batch all of these updates using [`Connection.executemany`](https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.executemany)

Reply:
I'm not sure it's the right use case here - since the query is slightly different in each case, and from what I understand with `executemany`, you want to provide a sequence of args against the same query.